vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java

```diff
@@ -90,9 +93,13 @@ public void close() {
     MongoCollection<JsonObject> coll = getCollection(collection, writeOption);
     Object id = document.getValue(ID_FIELD);
     if (id == null) {
-      coll.insertOne(document, convertCallback(resultHandler, wr -> document.getString(ID_FIELD)));
+      coll.insertOne(document, convertCallback(resultHandler, wr -> useObjectId ? document.getJsonObject(ID_FIELD).getString(OID_FIELD) : document.getString(ID_FIELD)));
     } else {
-      coll.replaceOne(wrap(new JsonObject().put(ID_FIELD, document.getValue(ID_FIELD))), document, convertCallback(resultHandler, result -> null));
+      JsonObject filter = new JsonObject();
+      encodeKeyWhenUseObjectId(document);
```

Comment:
Sadly, we can't do that within a lambda. "document" as to be final or effectively final. You can't assign anything back to it. It seems ridiculous since you can mutate the object anyway, like I am doing. If you know a way to make the assignment within a lambda, I would surely like to know it! It's the thing I dislike most about Java 8.

I return the JsonObject from the method in the event that someone else wants to use it fluently.